src/main/java/conflux/dex/service/blockchain/BlockchainSettlementService.java

```diff
@@ -201,13 +211,51 @@ private void settle(Settleable data) throws Exception {
 			// 2. service restarted and last item not settled on chain yet
 			this.validateNonce(data, false);
 			this.sendTransaction(data, context);
+			txHash = data.getSettledTxHash();
 		}
 		
 		data.updateSettlement(this.dao, SettlementStatus.OnChainSettled);
 		
 		this.monitor.add(data);
+
+		if (txHash != null) {
+			waitTxExecuted(data);
+		}
 	}
-	
+
+	private void waitTxExecuted(Settleable data) {
+		String txHash = data.getSettledTxHash();
+		while (true) {
+			// maybe it has been paused by tx monitor already.
+			if (this.healthService.getPauseSource().isPresent()) {
+				break;
+			}
+			if (TransactionRecorder.Error.TxDiscarded.equals(data.getRecorder().getLast().error)) {
+				// tx monitor may set the error to `TxDiscarded`, and fire an event in order to resubmit this tx.
+				break;
+			}
+			Optional<Receipt> opt;
+			try {
+				opt = this.blockchain.getAdmin().getCfx().getTransactionReceipt(txHash).sendAndGet();
+			} catch (RpcException e) {
+				logger.error("rpc exception while fetching tx receipt, tx {}", txHash, e);
+				opt = Optional.empty();
+			}
+			// not executed, or failed
+			if (!opt.isPresent() || opt.get().getOutcomeStatus() != 0) {
```

Comment:
what if tx failed?

Reply:
If the transaction fails, the transaction monitor will pause the system, 
 the next loop will break by checking pause source, do nothing here.